torch/csrc/jit/graph_executor.cpp

```diff
@@ -0,0 +1,396 @@
+      v = v.defined() ? autograd::Variable(v).data() : at::Tensor();
+    }
+  }
+  // inplace to avoid allocations
+  void wrapTensors(tensor_list & list) {
+    for(auto & v : list) {
+      v = autograd::make_variable(v);
+    }
+  }
+  // TODO: both captureInputs/captureOutputs have to iterate over the entire
+  // capture list because the list is not sorted with inputs before outputs.
+  // We cannot capture both, because we actually destroy our input lists (to release references)
+  // before we get the outputs list.
+  void captureInputs(ExecutionPlanAutogradFunction & grad_fn, variable_tensor_list & inputs) {
+    auto & capture_desc = grad.df_input_captures;
+    size_t N = capture_desc.size();
+    for(size_t i = 0; i < N; ++i) {
+      if(capture_desc[i].kind == Capture::Kind::Input) {
+        size_t offset = capture_desc[i].offset;
+        grad_fn.captures[i] = autograd::SavedVariable(autograd::Variable(inputs[offset]), false);
+      }
+    }
+  }
+  void captureOutputs(ExecutionPlanAutogradFunction & grad_fn, variable_tensor_list & outputs) {
+    auto & capture_desc = grad.df_input_captures;
+    size_t N = capture_desc.size();
+    for(size_t i = 0; i < N; ++i) {
+      if(capture_desc[i].kind == Capture::Kind::Output) {
+        size_t offset = capture_desc[i].offset;
+        grad_fn.captures[i] = autograd::SavedVariable(autograd::Variable(outputs[offset]), true);
+      }
+    }
+  }
+
+  variable_tensor_list runWithGrad(variable_tensor_list inputs) {
+    auto grad_fn = std::make_shared<ExecutionPlanAutogradFunction>(grad_executor, grad.df_input_captures.size());
+    // hook up the outputs of df to the gradient functions of the inputs that require
+    // gradients
+    for(auto idx : grad.df_output_vjps) {
+      autograd::Variable v(inputs[idx]);
+      // TODO: this kinda stuff is _way_ to low level to the public API of variable.
+      // Why do I have to care here whether v has a grad_fn or grad accumulator?
+      // Why do I have to care here about output_nr? I just want to say
+      // grad_fn->setOutputTo(i, v.input_port());
+      grad_fn->next_functions.emplace_back(v.grad_fn() ? v.grad_fn() : v.grad_accumulator(), v.output_nr());
+    }
+    captureInputs(*grad_fn, inputs);
+
+    unwrapVariables(inputs);
+    tensor_list outputs;
+    InterpreterState(f).runOneStage(std::move(inputs), outputs);
+    wrapTensors(outputs);
+    // hookup the gradients for the output tensors that require gradients
+    // to the inputs to our gradient function df
+    // TODO - XXX - if any output is the same tensor multiple times, views have to be
+    // setup here. We need to refactor autograd until it is safe for
+    // tensors to be constructed without all the viewing infrastructure.
+    // this is currently intentionally not done here so we can get an idea of our
+    // perf before introducing overhead for correctness
+    for(auto idx : grad.df_input_vjps) {
+      autograd::Variable o(outputs[idx]);
```

Comment:
please please please please use `static_cast` for this

Reply:
I was making this change, but I am now confused. The comment on the constructor says:

```
 // Implicitly casts a Tensor to a Variable. This should only be called on
 // Tensors which you know are actually Variables.
```

What use are these constructors except for the exact situation where I am using them? If I shouldn't be using them, then should we delete them?

Reply:
No, if I understand correctly, the main purpose of these constructors is that C++ can do implicit conversions when you're passing an `at::Tensor` to a function that takes in a `Variable` (yup, unsafe). However everywhere in the code I've seen `static_casts` and they describe the intent more explicitly